components/chat/useChatSocket.tsx

```diff
@@ -42,116 +47,146 @@ const useChatSocket = (isEmpty?: boolean) => {
   socketRef.current = socket;
+  latestAgentMessageRef.current = latestAgentMessage;
 
   const handleError = useCallback((error: string) => {
     setGenerating(false);
     setWaitingForUserResponse(false);
     setError(error);
     setMessages((prevMessages) => {
-      const updatedMessages = [...prevMessages];
-      if (latestAgentMessageIndex.current !== -1) {
-        // Append the error to the latest message
-        updatedMessages[latestAgentMessageIndex.current].error = `${error}`;
-      } else {
-        // If there are no previous messages, create a new error message
-        updatedMessages.push({
-          type: MessageType.Agent,
-          name: '',
-          message:
-            'The script encountered an error. You can either restart the script or try to continue chatting.',
-          error,
-        });
+      if (!latestAgentMessageRef.current.type) {
+        return [
+          ...prevMessages,
+          {
+            type: MessageType.Agent,
+            name: '',
+            message:
+              'The script encountered an error. You can either restart the script or try to continue chatting.',
+            error,
+          },
+        ];
       }
-      return updatedMessages;
+
+      latestAgentMessage.error = `${error}`;
+      setLatestAgentMessage({} as Message);
+      return [...prevMessages, { ...latestAgentMessage }];
     });
   }, []);
 
   // handles progress being received from the server (callProgress style frames).
   const handleProgress = useCallback(
-    ({
-      frame,
-      state,
-      name,
-    }: {
-      frame: CallFrame;
-      state: Record<string, CallFrame>;
-      name?: string;
-    }) => {
-      if (!frame.error && frame.toolCategory === 'provider') {
-        return;
+    ({ frame, name }: { frame: Frame; name?: string }) => {
+      workQueue.current.push({ frame, name });
+    },
+    []
+  );
+
+  // Function to process only the number of messages that were in the queue when processing started
+  const processWorkQueue = useCallback(() => {
+    const initialQueueLength = workQueue.current.length;
+    if (initialQueueLength === 0) return;
+
+    for (let i = 0; i < initialQueueLength; i++) {
+      if (workQueue.current.length === 0) break; // Stop if the queue is empty
+
+      const { frame, name } = workQueue.current.shift()!; // Remove and process the first message in the queue
+
+      if (!latestAgentMessageRef.current.type)
+        latestAgentMessageRef.current.type = MessageType.Agent;
+      if (!latestAgentMessageRef.current.name)
+        latestAgentMessageRef.current.name = name;
+      if (latestAgentMessageRef.current.component)
+        latestAgentMessageRef.current.component = null;
+
+      if (!frame.type.startsWith('call')) {
+        if (frame.type === 'runStart') {
+          latestAgentMessageRef.current.message =
+            'Waiting for model response...';
+          setLatestAgentMessage({ ...latestAgentMessageRef.current });
```

Comment:
@tylerslaton This is where the initial message is added back.

Reply:
Running this locally this is what I'm seeing. The "Waiting for model response..." was guaranteed to be instant. For whatever reason I'm not seeing what you linked as instant.


https://github.com/user-attachments/assets/01c7ea38-e4db-4b97-9f6f-ae2bafeda5a5

Reply:
That is only for the first message and is still a much smaller wait than what we were seeing before. I will see if I can speed this up somehow.